src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs

```diff
@@ -0,0 +1,2164 @@
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_proxy);
+            }
+
+            if (_cookieUsePolicy == CookieUsePolicy.UseSpecifiedCookieContainer &&
+                _cookieContainer == null)
+            {
+                throw new InvalidOperationException(SR.net_http_invalid_cookiecontainer);
+            }
+
+            CheckDisposed();
+
+            SetOperationStarted();
+
+            TaskCompletionSource<HttpResponseMessage> tcs = new TaskCompletionSource<HttpResponseMessage>();
+
+            // Create RequestState object and save current values of handler settings.
+            RequestState state = new RequestState();
+            state.Tcs = tcs;
+            state.CancellationToken = cancellationToken;
+            state.RequestMessage = request;
+            state.Handler = this;
+            state.CheckCertificateRevocationList = _checkCertificateRevocationList;
+            state.ServerCertificateValidationCallback = _serverCertificateValidationCallback;
+            state.WindowsProxyUsePolicy = this.WindowsProxyUsePolicy;
+            state.Proxy = this.Proxy;
+            state.ServerCredentials = _serverCredentials;
+
+            try
+            {
+                Task.Factory.StartNew(StartRequest, state);
+            }
+            catch (Exception ex)
+            {
+                throw new HttpRequestException("SendAsync", ex);
+            }
+
+            return tcs.Task;
+        }
+
+        private static uint ChooseAuthScheme(uint supportedSchemes)
+        {
+            foreach (uint authScheme in s_authSchemePriorityOrder)
+            {
+                if ((supportedSchemes & authScheme) != 0)
+                {
+                    return authScheme;
+                }
+            }
+
+            return 0;
+        }
+
+        private static void WinHttpStatusCallback(
+            IntPtr handle,
+            IntPtr context,
+            uint internetStatus,
+            IntPtr statusInformation,
+            uint statusInformationLength)
+        {
+            RequestState state;
+
+            // TODO: Use logging.
+
+            if (internetStatus != Interop.WinHttp.WINHTTP_CALLBACK_STATUS_REDIRECT &&
+                internetStatus != Interop.WinHttp.WINHTTP_CALLBACK_STATUS_SENDING_REQUEST)
+            {
+                return;
+            }
+
+            // Get request state object.
+            try
+            {
+                GCHandle gch = GCHandle.FromIntPtr(context);
+                state = (RequestState)gch.Target;
+            }
+            catch (Exception)
+            {
+                // TODO: We should log this but an exception here is pretty rare and
+                // is probably the result of a programming error.
+
+                // Closing the WinHTTP request handle will cancel the operation.
+                Interop.WinHttp.WinHttpCloseHandle(handle);
```

Comment:
There are a lot of calls to WinHttpCloseHandle.  That makes me a bit twitchy, can a SafeHandle be added for goodness?